PurchasesTests/Attribution/AttributionFetcherTests.swift

```diff
@@ -0,0 +1,235 @@
+
+        attributionFetcher.postAttributionData(["something": "else"],
+                                               from: .adjust,
+                                               forNetworkUserId: userID)
+
+        expect(self.backend.invokedPostAttributionDataCount) == 1
+
+    }
+
+    func testPostAttributionDataDoesntSkipIfNetworkUserIdChanged() {
+        let userID = "userID"
+        backend.stubbedPostAttributionDataCompletionResult = (nil, ())
+
+        attributionFetcher.postAttributionData(["something": "here"],
+                                               from: .adjust,
+                                               forNetworkUserId: userID)
+        expect(self.backend.invokedPostAttributionDataCount) == 1
+
+        attributionFetcher.postAttributionData(["something": "else"],
+                                               from: .facebook,
+                                               forNetworkUserId: userID)
+
+        expect(self.backend.invokedPostAttributionDataCount) == 2
+    }
+
+    func testPostAttributionDataDoesntSkipIfSameUserIdButDifferentNetwork() {
+        backend.stubbedPostAttributionDataCompletionResult = (nil, ())
+
+        attributionFetcher.postAttributionData(["something": "here"],
+                                               from: .adjust,
+                                               forNetworkUserId: "attributionUser1")
+        expect(self.backend.invokedPostAttributionDataCount) == 1
+
+        attributionFetcher.postAttributionData(["something": "else"],
+                                               from: .facebook,
+                                               forNetworkUserId: "attributionUser2")
+
+        expect(self.backend.invokedPostAttributionDataCount) == 2
+    }
+
+    func testPostAppleSearchAdsAttributionIfNeededSkipsIfATTFrameworkNotIncludedOnNewOS() {
+        if #available(iOS 14, macOS 11, tvOS 14, *) {
+            systemInfo.stubbedIsOperatingSystemAtLeastVersion = true
+            MockAttributionTypeFactory.shouldReturnAdClientClass = true
+            MockAttributionTypeFactory.shouldReturnTrackingManagerClass = false
+
+            self.attributionFetcher.postAppleSearchAdsAttributionIfNeeded()
+
+            expect(MockAdClient.requestAttributionDetailsCallCount) == 0
+        }
+    }
+
+    func testPostAppleSearchAdsAttributionIfNeededPostsIfATTFrameworkNotIncludedOnOldOS() {
+        if #available(iOS 14, macOS 11, tvOS 14, *) {
+            systemInfo.stubbedIsOperatingSystemAtLeastVersion = false
+            MockAttributionTypeFactory.shouldReturnAdClientClass = true
+            MockAttributionTypeFactory.shouldReturnTrackingManagerClass = false
+
+            self.attributionFetcher.postAppleSearchAdsAttributionIfNeeded()
+
+            expect(MockAdClient.requestAttributionDetailsCallCount) == 1
+        }
+    }
+
+    func testPostAppleSearchAdsAttributionIfNeededSkipsIfIAdFrameworkNotIncluded() {
```

Comment:
Whould we also test for new OS and for old OS?

Reply:
I didn't think it was worth the trouble, since the logic for the iAd framework remains the same for old or new OS versions, so it'd either break for both or work for both